services/offline/src/lib/dhis2-connection-status.tsx

```diff
@@ -0,0 +1,201 @@
+}): JSX.Element => {
+    const [isConnected, setIsConnected] = useState(true)
+    const offlineInterface = useOfflineInterface()
+    const engine = useDataEngine()
+
+    const smartIntervalRef = useRef(null as null | SmartInterval)
+
+    /**
+     * Update state and potentially reset ping backoff and update
+     * the lastConnected value in localStorage
+     */
+    const updateConnectedState = useCallback((newIsConnected) => {
+        // use 'set' with a function as param to get latest isConnected
+        // without needing it as a dependency for useCallback
+        setIsConnected((prevIsConnected) => {
+            // if value changed, reset ping backoff to initial
+            if (newIsConnected !== prevIsConnected) {
+                smartIntervalRef.current?.resetDelayToInitial()
+            }
+            // if disconnected and EITHER 1. coming from connected or
+            // 2. there is no last-connect val, update the val in localStorage
+            if (!newIsConnected && (prevIsConnected || !getLastConnected())) {
+                updateLastConnected()
+            }
+            return newIsConnected
+        })
+    }, [])
+
+    // Note that the SW is configured to not cache ping requests and won't
+    // trigger `handleChange` below to avoid redundant signals. This also
+    // helps to detect the connectivity status when the SW is not available
+    // for some reason (maybe private browsing, first installation, or
+    // insecure browser context)
+    const ping = useCallback(() => {
+        return engine
+            .query(pingQuery)
+            .catch((err) => {
+                if (/Unexpected token 'p'/.test(err.message)) {
+                    // The request succeeded; this is just a weird endpoint
+                    // (rest of the error is '"pong" is not valid JSON')
+                    // todo: maybe change link handling of this endpoint
+                    return
+                } else {
+                    // It's a different error; throw to the next catch handler
+                    throw err
+                }
+            })
+            .then(() => {
+                // Ping is successful; set 'connected'
+                updateConnectedState(true)
+            })
+            .catch((err) => {
+                // Can get here if unauthorized, network error, etc.
+                console.error('Ping failed:', err.message)
+
+                // Only network errors should change status
+                const isNetworkErr = /^An unknown network error occurred$/.test(
+                    err.message
+                )
+                if (isNetworkErr) {
+                    updateConnectedState(false)
+                }
+            })
+    }, [engine, updateConnectedState])
+
+    /** Called when SW reports updates from incidental network traffic */
+    const onUpdate = useCallback(
+        ({ isConnected: newIsConnected }) => {
+            console.log('handling update from sw')
```

Comment:
Remove console.log
```suggestion
```

Reply:
This log (and all the others) I find pretty useful for testing this feature out thoroughly, because it's sometimes hard to inspect what's going on with the SW, pings, smart interval, etc. otherwise -- my plan is to remove them all once the PR is done being reviewed, but I'll add a todo above this line to remind me

Reply:
OK, but since this is a PR against master and I feel it's in its final stages we should now either get rid of them or keep them in but ensure they only show on the console in development mode. I.e. sth like this:
```js
function devLog() {
    if (process.env === 'development') {
        console.log(arguments)
    }
}
```

Reply:
Added this to keep them available if wanted, but otherwise omitted: https://github.com/dhis2/app-runtime/pull/1203/commits/0389e3f25836d271f200396a7e2345f74085d64f

They're super chatty; I doubt anyone would want to see them when they're working on an app that uses the hook